### PR TITLE
Introduce functional adt for chain handling.

### PR DESCRIPTION
**Pull request by happi**
Major rewrite of chain handling.

The chain is now kept in a functional ADT handled by aec_chain_state.erl
This chain is kept by a server (aec_chain_server.erl) and the public interface is still aec_chain.erl.
The interface is mostly backwards compatible with the old implementation but some internal functions that were exposed (for testing and for the chain_state) are deprecated and removed.
The new chain can handle alternative chains and blocks without parents.
The state is check pointed at each block on the main chain. (This will be cleaned up by a checkpoint GC in the future.)
The chain automatically switches to the chain with most work.

Closing #64 and #59

### CODE REVIEW

**Comment by gorillainduction** [comment]
Sorry for rewriting so much code, guys. This is not an easy PR to review.

File: apps/aecore/src/aec_chain_state.erl [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+                true  -> State;
+                false -> internal_error({different_header_existing_block, Old, Node})
+            end
+    end.
+
+check_update_after_insert(Node, State) ->
+    case determine_chain_relation(Node, State) of
+        off_chain -> State;
+        in_chain -> update_state_tree(Node, State);
+        {fork, ForkNode} ->
```
Comment (sennui) [inline apps/aecore/src/aec_chain_state.erl]: please confirm: here we switch over to fork always when we get it.

We need to add difficulty check for incoming fork (just to make sure if someone is not cheating with lower target). Actually, we need to compute total difficulty in each branch from fork point

Comment (happi) [reply apps/aecore/src/aec_chain_state.erl]: Yes, this is done. Look at determine_new_header_top.

File: apps/aecore/src/aec_chain_server.erl [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    process_flag(trap_exit, true),
+    GenesisHeader = aec_blocks:to_header(GenesisBlock),
+    {ok, S2} = insert_header(GenesisHeader, new_state()),
+    {ok, S3} = insert_block(GenesisBlock, S2),
+    {ok, S3}.
+
+
+%% State preserving functions
+handle_call(top, _From, State) ->
+    {reply, top(State), State};
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_server.erl]: Recognizing that this implementation is a 'shadow' version of `aec_chain`, is the intention to rename this module as `aec_chain`? It's not common practice to have API functions in one module do direct gen_server calls to a server implemented in another. If API functions are added to this module later, it seems as if the intuitive names for these functions are already reserved for internal functions (though perhaps with different arity). I would suggest using an 'internal naming' scheme for these – either `*_int()` or just `*_()` (which is a convention e.g. Tony Rogvall likes to use.)

File: apps/aecore/src/aec_chain_state.erl [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+set_top_header_hash(H, State) when is_binary(H) -> State#{top_header_hash => H}.
+
+get_top_block_hash(#{top_block_hash := H}) -> H.
+set_top_block_hash(H, State) when is_binary(H) -> State#{top_block_hash => H}.
+
+%%%-------------------------------------------------------------------
+%%% Internal ADT for differing between blocks and headers
+%%%-------------------------------------------------------------------
+
+-record(node, { type    :: 'block' | 'header'
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_state.erl]: For persistence reasons, would it perhaps be preferable to use maps here?

File: apps/aecore/src/aec_chain_state.erl [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    Height   = node_height(Node),
+    Hash     = Node#node.hash,
+    PrevHash = prev_hash(Node),
+    assert_previous_height(Height, PrevHash, State),
+    case get_top_header_hash(State) of
+        PrevHash -> new_top;
+        undefined when Height > 0 ->
+            off_chain; %% No proper chain yet.
+        undefined when Height =:= 0 ->
+            new_top; %% A genesis block. TODO: Should this be checked?
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_state.erl]: I think it should be checked, yes.

File: apps/aecore/src/aec_chain_state.erl [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        Hash ->
+            in_chain; %% This is the top header
+        TopHash when is_binary(TopHash), Height =:= 0 ->
+            %% A new genesis block. TODO: This
+            error(new_genesis_block_nyi);
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_state.erl]: Why not `internal_error()`?

File: apps/aecore/src/aec_chain_state.erl [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+    end.
+
+find_new_header_top_from_node(Node, State) ->
+    Height = node_height(Node),
+    Hash = hash(Node),
+    case children(blocks_db_find_at_height(Height + 1, State), Hash) of
+        [] -> [Node];
+        [NextNode] -> find_new_header_top_from_node(NextNode, State);
+        [_|_]  = List ->
+            lists:flatten([find_new_header_top_from_node(N, State)
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_state.erl]: why not `lists:flatmap(fun(N) -> find_new_header_top_from_node(N, State) end, List)`?

File: apps/aecore/src/aec_chain_state.erl [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+children([N|Ns], ParentHash, Acc) ->
+    case prev_hash(N) =:= ParentHash of
+	true  -> children(Ns, ParentHash, [N|Acc]);
+	false -> children(Ns, ParentHash, Acc)
+    end.
+
+determine_new_header_top(Nodes, State) ->
+    Difficulties = [{total_difficulty_at_hash(N#node.hash, State), N}
+                    || N <- Nodes],
+    {_D, N}  = lists:last(lists:sort(Difficulties)),
```
Comment (uwiger) [inline apps/aecore/src/aec_chain_state.erl]: Why not `lists:max(Difficulties)`?

File: apps/aecore/test/aec_chain_tests.erl [hunk 8]
```diff
@@ -269,15 +255,14 @@ get_work_test_() ->
                BH0 = aec_blocks:to_header(B0),
                ?assertEqual({ok, BH0}, aec_chain:top_header()),
 
-               %% Check difficulty of genesis - for readability of the test.
-               1.0 = aec_headers:difficulty(BH0),
+               %% Check difficulty of genesis
+	       %% NOTE: floating point matching... not safe.
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: My guess is it's *pretty* safe in this case, but still, if it's regarded as unsafe, wouldn't it be better to change it? E.g. into something like `trunc(1000 * ?GENESIS_DIFFICULTY) = trunc(1000, ...)`

File: apps/aecore/test/aec_chain_tests.erl [hunk 9]
```diff
@@ -287,7 +272,8 @@ get_work_test_() ->
                ?assertEqual({ok, BH0}, aec_chain:top_header()),
 
                %% Check difficulty of genesis - for readability of the test.
-               1.0 = aec_headers:difficulty(BH0),
+	       %% NOTE: floating point matching... not safe.
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above.

File: apps/aecore/test/aec_chain_tests.erl [hunk 10]
```diff
@@ -671,7 +597,7 @@ longest_header_chain_test_() ->
                %% Generate the two header chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe')

File: apps/aecore/test/aec_chain_tests.erl [hunk 11]
```diff
@@ -688,46 +614,18 @@ longest_header_chain_test_() ->
...
-                 lists:nthtail(1, MainHC))
+               ?assertEqual({ok, HA3}, aec_chain:top_header())
        end},
       {"The alternative header chain has more work - case alternative chain is higher. Force chain including genesis.",
        fun() ->
                %% Generate the two header chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe')

File: apps/aecore/test/aec_chain_tests.erl [hunk 12]
```diff
@@ -744,45 +642,18 @@ longest_header_chain_test_() ->
...
-                 lists:nthtail(1, MainHC))
+               ?assertEqual({ok, HA3}, aec_chain:top_header())
        end},
       {"The alternative header chain has more work - case alternative chain is less high",
        fun() ->
                %% Generate the two header chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 13]
```diff
@@ -841,12 +686,12 @@ longest_header_chain_test_() ->
...
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 14]
```diff
@@ -913,21 +739,18 @@ longest_block_chain_test_() ->
                %% Generate the two block chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 15]
```diff
@@ -985,47 +786,28 @@ longest_block_chain_test_() ->
...
+                                                      aec_blocks:trees(BA3)))
+
        end},
       {"The alternative block chain has more work - case both main and alternative chains with all blocks. Two blocks in common.",
        fun() ->
                %% Generate the two block chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 16]
```diff
@@ -1092,45 +852,28 @@ longest_block_chain_test_() ->
...
-                 end,
-                 lists:nthtail(2, MainHC))
        end},
       {"The alternative block chain has more work - case main chain with only block corresponding to top header (not contiguous to genesis), and alternative chain with only block corresponding to top header (not contiguous to genesis). Only genesis block in common.",
        fun() ->
                %% Generate the two block chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 17]
```diff
@@ -1183,41 +910,24 @@ longest_block_chain_test_() ->
...
-                 end,
-                 lists:nthtail(1, MainHC))
        end},
       {"The alternative block chain has more work - case main chain with only one block contiguous to genesis, and alternative chain with two blocks contiguous to genesis. Only genesis block in common.",
        fun() ->
                %% Generate the two block chains.
                B0 = fake_genesis_block_with_difficulty(),
                0 = aec_blocks:height(B0), %% For readability of the test.
-               1.0 = aec_blocks:difficulty(B0), %% For readability of the test.
+               ?GENESIS_DIFFICULTY = aec_blocks:difficulty(B0),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Same comment as above ('not safe').

File: apps/aecore/test/aec_chain_tests.erl [hunk 18]
```diff
@@ -1274,30 +966,17 @@ longest_block_chain_test_() ->
 
                %% Check top block changed.
                ?assertEqual({ok, HA3}, aec_chain:top_header()),
-               ?assertEqual({ok, BA2}, aec_chain:top_block()),
-
+	       
+	       %% !!!! Race condition here !!!!
+	       timer:sleep(2000),
```
Comment (uwiger) [inline apps/aecore/test/aec_chain_tests.erl]: Isn't there any way to achieve a sync point here without resorting to timers? A TODO perhaps?

File: apps/aecore/src/aec_chain_server.erl [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    gen_server:stop(?SERVER).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+ 
+init(_Args = [GenesisBlock]) ->
+    process_flag(trap_exit, true),
+    GenesisHeader = aec_blocks:to_header(GenesisBlock),
+    {ok, S2} = insert_header(GenesisHeader, new_state()),
```
[the remark this answers is not in the document]
Comment (happi) [reply apps/aecore/src/aec_chain_server.erl]: Good catch, refactoring artifact.

File: apps/aecore/src/aec_chain_server.erl (hunk 2 above)
Comment (happi) [reply apps/aecore/src/aec_chain_server.erl]: I think it is a better and cleaner practice to separate the outside API and the server implementation. We can discuss offline, and then come back here for a conclusion.

File: apps/aecore/src/aec_chain_state.erl (hunk 4 above)
Comment (happi) [reply apps/aecore/src/aec_chain_state.erl]: This is only when there was no genesis block before.

File: apps/aecore/test/aec_chain_tests.erl (hunk 18 above)
Comment (happi) [reply apps/aecore/test/aec_chain_tests.erl]: This timer is not needed with the new implementation which only uses one server.
will remove.

File: apps/aecore/src/aec_chain_state.erl (hunk 7 above)
Comment (gorillainduction) [reply apps/aecore/src/aec_chain_state.erl]: Good one. I will update.

File: apps/aecore/test/aec_chain_tests.erl (hunk 18 above)
Comment (gorillainduction) [reply apps/aecore/test/aec_chain_tests.erl]: The race is not there anymore. It was an artifact in the old version. I remove this.

**Comment by gorillainduction** [comment]
Please note that we have aimed for correctness and not speed/space optimizations. The plan is to introduce such optimizations in the next step.

File: apps/aecore/src/aec_chain_state.erl (hunk 6 above)
Comment (gorillainduction) [reply apps/aecore/src/aec_chain_state.erl]: Sure.

File: apps/aecore/src/aec_chain_state.erl (hunk 1 above)
Comment (gorillainduction) [reply apps/aecore/src/aec_chain_state.erl]: I will make this more clear in the code.

File: apps/aecore/src/aec_chain_state.erl (hunk 3 above)
Comment (happi) [reply apps/aecore/src/aec_chain_state.erl]: Maybe. It depends on how we will do persistence and how we will handle code change. If we use mnesia or dets a record is probably easiest. For RocksDB or something else we would need to do some serialization and deserialization anyway. Records might be slightly more efficient here than records. Versioning of the data will have to be handled anyway.

I will look at persistence next and perhaps come back and change this if needed but for now I suggest we go with records.

**Pull request merged**